storage/innobase/buf/buf0dblwr.cc

```diff
@@ -602,20 +602,58 @@ static void buf_dblwr_check_block(const buf_page_t *bpage) noexcept
 }
 #endif /* UNIV_DEBUG */
 
+ATTRIBUTE_COLD void buf_dblwr_t::print_info() const noexcept
+{
+  mysql_mutex_assert_owner(&mutex);
+  const slot *flush_slot= active_slot == &slots[0] ? &slots[1] : &slots[0];
+
+  sql_print_information("Innodb: Double Write State\n"
+      "-------------------\n"
+      "Batch running : %s\n"
+      "Active Slot - first_free: %lu reserved:  %lu\n"
+      "Flush  Slot - first_free: %lu reserved:  %lu\n"
+      "-------------------",
+      (batch_running ? "true" : "false"),
+      active_slot->first_free, active_slot->reserved,
+      flush_slot->first_free, flush_slot->reserved);
+}
+
 bool buf_dblwr_t::flush_buffered_writes(const ulint size) noexcept
 {
   mysql_mutex_assert_owner(&mutex);
   ut_ad(size == block_size());
 
-  for (;;)
+  const int max_count= 60 * 60;
+  const int first_log_count= 30;
+  int log_count= first_log_count;
+
+  for (int count= 0;;)
   {
     if (!active_slot->first_free)
       return false;
     if (!batch_running)
       break;
-    my_cond_wait(&cond, &mutex.m_mutex);
-  }
 
+    timespec abstime;
+    set_timespec(abstime, 1);
+    my_cond_timedwait(&cond, &mutex.m_mutex, &abstime);
+
+    if (++count < first_log_count && !(count % 5))
+    {
+      sql_print_information(
+        "Long wait (%d seconds) for double-write buffer flush.", count);
+      buf_pool.print_flush_info();
+      print_info();
+    }
+    else if (!(count % log_count))
+    {
+      sql_print_warning(
+        "Long wait (%d seconds) for double-write buffer flush.", count);
```

Comment:
I don’t think that the common formatting style allows line breaks after `(`.

Could we make this output dependent on `innodb_fatal_semaphore_wait_threshold` and also kill the server when that limit is exceeded?